File: clojure/run-length-encoding/src/run_length_encoding.clj
(ns run-length-encoding (:require [clojure.string :as str]))

(defn run-length-encode
  "Encodes a string with run-length encoding."
  [plaintext]
  (->>
    plaintext
    (partition-by identity)
    (map #(str (if (= 1 (count %)) "" (count %)) (first %)))
    (str/join)))

(defn run-length-decode
  "Decodes a run-length-encoded string."
  [ciphertext]
  (->> 
    ciphertext
    (re-seq #"(\d*)(\D)")
    (mapcat (fn [[_ n ch]] (repeat (if (str/blank? n) 1 (parse-long n)) ch)))
    (str/join)))

File: python/run-length-encoding/run_length_encoding.py
"""Implement Run Length Encoding and Decoding"""

from itertools import groupby
import re

def decode(string):
    pairs = re.findall(r'(\d+)?(\D)', string)

    return ''.join(
        (int(count) if count else 1) * char
        for count, char in pairs
    )
    

def encode(string):
    def encode_group(letter, chars):
        count = sum(1 for _ in chars)
        return (str(count) if count > 1 else '') + letter

    return ''.join(
        encode_group(letter, chars)
        for letter, chars in groupby(string)
    )
    